### PR TITLE
#3332 create view rr modal

### LINKED ISSUE #3332: [Finance Redesign] - create view RR modal
### Description

We need to be able to display the information in an RR in the side page popup component. The timeline does not need to be part of the proposed solution. Look at the current reimbursement request view code to understand how it works and what to change.

### Acceptance Criteria

Looks like mock except timeline

### Proposed Solution

A view modal in the side page popup component

### Mocks

Image
The Actions should the same as they are for currently viewing a reimbursement request.

Image
The columns for products will should adjust to the context of the reimbursement (if there's one refund source and one project, there should only be item and costs; if there's multiple refund sources and one project, there should be columns for item, cost, and divide between budget and cash; if there's one refund source and multiple projects, there should be columns for item, cost, and project; if there's multiple refund sources and projects, there should be columns for item, cost, project, and divide between budget and cash)

### PR DESCRIPTION
**Pull request by JoshuaGoldberg**
## Changes

- Added a side page that pops up when selecting reimbursement requests.
- Page includes information about the request in an easy to understand format. 
- Also displays information for individual products, alongside receipts.
- Product information changes based on the quantity of refund sources/projects. If multiple projects exist, a project column is added. If both cash and budget are used, a breakdown of costs between the two sources is added.
- Made the design look as close to the mock as possible (had to change up a few symbols).
- Actions menu accessible in the top left for modifying/editing the request.
- Updating status via the dropdown refreshes the table to display the new data, without needing to refresh the page.

- In addition to the table, also added refund sources to hold refund data, allowing for reimbursement products to have multiple refund sources at once.
- Made changes to the create RR modal to reflect this new functionality (would be nice to further update this in the future)
- Added Timeline to the side panel (from a previous merge).
- Also fixed the editing modal to work with the new model for reimbursements, and made sure they submit/update properly.

Side note: reverted the prior view route edit I made since it no longer conflicts, as views can simply be redirected to the reimbursement table page now.

## Notes

~~Backend changes were a little messy since I was confused initially, but seem to all work fine. My main concern would be that the cashAmount and ```budgetAmount fields were added to too many of the reimbursement request types, but if so, that's a pretty simple fix. Figured it would be better to store a little extra information than to not store enough.~~

Changed to support refundsources[] instead. My main concern now is that I haven't had too much time to touch the create RR modal itself, and currently it still technically only supports two different refund sources (with the categories being hardcoded in as elements of refund sources). Obviously, it would be nicer to have it be more flexible with any number of refund sources. Thankfully, this should be isolated to just the create RR modal logic, and everything else should work regardless. 

## Test Cases

- Case A
- Edge case
- ...

## Screenshots
Using both refund sources and one project:
image

Using one refund source and multiple projects
image

Small window
image

## To Do

Just what was noted above, potentially clean up some stuff getting stored, in case its not needed. 

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #3332

### CODE REVIEW

File: src/backend/src/prisma/migrations/20250402021534_finance_redesign/migration.sql [hunk 1]
```diff
@@ -256,4 +256,23 @@ ALTER TABLE "Reimbursement_Request_Comment" ADD CONSTRAINT "Reimbursement_Reques
...
+    "refundSourceId" TEXT NOT NULL,
+    "indexCodeId" TEXT NOT NULL,
+    "amount" INTEGER NOT NULL,
+    "reimbursementProductId" TEXT,
+
+    CONSTRAINT "RefundSource_pkey" PRIMARY KEY ("refundSourceId")
+);
+
+-- CreateIndex
+CREATE UNIQUE INDEX "RefundSource_indexCodeId_key" ON "RefundSource"("indexCodeId");
```
Comment (wavehassman) [inline src/backend/src/prisma/migrations/20250402021534_finance_redesign/migration.sql]: put everything in one migration, so just delete this line and then delete the second migration

File: src/backend/src/prisma/schema.prisma [hunk 2]
```diff
@@ -625,6 +625,16 @@ model Reimbursement_Product {
   reimbursementProductReason   Reimbursement_Product_Reason @relation(fields: [reimbursementProductReasonId], references: [reimbursementProductReasonId])
   reimbursementRequestId       String
   reimbursementRequest         Reimbursement_Request        @relation(fields: [reimbursementRequestId], references: [reimbursementRequestId])
+  refundSources                RefundSource[]
+}
+
+model RefundSource {
```
Comment (wavehassman) [inline src/backend/src/prisma/schema.prisma]: this is my fault, but put an underscore between Refund and Source to match the other models

File: src/frontend/src/apis/finance.api.ts [hunk 3]
```diff
@@ -68,6 +68,7 @@ export const uploadSingleReceipt = (file: File, id: string) => {
  * @returns the created reimbursement request
  */
 export const createReimbursementRequest = (formData: CreateReimbursementRequestPayload) => {
+  console.log(formData);
```
Comment (wavehassman) [inline src/frontend/src/apis/finance.api.ts]: you can take this out

File: src/frontend/src/pages/FinancePage/ReimbursementRequestDetailPage/ReimbursementProductsView.tsx [hunk 4]
```diff
@@ -15,26 +15,65 @@ const ReimbursementProductsView: React.FC<ReimbursementRequestProductsViewProps>
...
+                  {uniqueWbsElementsWithProducts
+                    .get(key)
+                    ?.map((product, index) => <div key={index}>${centsToDollar(product.cost)}</div>)}
                 </TableCell>
+                {multipleRefundSources && (
+                  <TableCell>
+                    {uniqueWbsElementsWithProducts
+                      .get(key)
+                      ?.map((product, index) => (
+                        <div key={index}>{product.refundSources.map((rs) => `$${centsToDollar(rs.amount)}`).join('/')}</div>
```
Comment (wavehassman) [inline src/frontend/src/pages/FinancePage/ReimbursementRequestDetailPage/ReimbursementProductsView.tsx]: for reimbursement #1 this is suppose to be $200 budget and $800 cash, but rn show up at $8 and $2, use the dollarPipe

File: src/frontend/src/pages/FinancePage/ReimbursementRequestDetailPage/ReimbursementRequestDetailsView.tsx [hunk 5]
```diff
@@ -477,60 +464,139 @@ const ReimbursementRequestDetailsView: React.FC<ReimbursementRequestDetailsViewP
...
   const statusTypes = sortedStatus.map((status) => status.type);
   const recentStatus = statusTypes[statusTypes.length - 1];
-  return (
-    <PageLayout
-      title={`Reimbursement Request #${reimbursementRequest.identifier} (${fullNamePipe(reimbursementRequest.recipient)})`}
-      chips={
-        <Box id="status" display="flex">
-          {statusTypes.length > 0 && <ReimbursementRequestStatusPill status={recentStatus} />}
+
+  const getStatusColor = (status: string): string => {
```
Comment (wavehassman) [inline src/frontend/src/pages/FinancePage/ReimbursementRequestDetailPage/ReimbursementRequestDetailsView.tsx]: Could you not just use the ReimbursementRequestStatusPill?

File: src/frontend/src/pages/FinancePage/ReimbursementRequestDetailPage/ReimbursementRequestDetailsView.tsx [hunk 6]
```diff
@@ -477,60 +464,139 @@ const ReimbursementRequestDetailsView: React.FC<ReimbursementRequestDetailsViewP
...
-        {
-          name: 'Finance',
-          route: routes.FINANCE
+      )
+    },
+    { content: fullNamePipe(reimbursementRequest.recipient) },
+    {
+      content: keys.map((key) => displayEnum(key)).join(', ')
+    },
+    { content: `$${centsToDollar(reimbursementRequest.totalCost)}` },
```
Comment (wavehassman) [inline src/frontend/src/pages/FinancePage/ReimbursementRequestDetailPage/ReimbursementRequestDetailsView.tsx:478]: Again use dollarsPipe, this may also come up in the table. I would just go back and make sure all the costs match up with the seed data

File: src/frontend/src/pages/FinancePage/ReimbursementRequestDetailPage/ReimbursementRequestDetailsView.tsx [hunk 7]
```diff
@@ -477,60 +464,139 @@ const ReimbursementRequestDetailsView: React.FC<ReimbursementRequestDetailsViewP
...
+    { label: 'SABO Number', icon: <SellIcon fontSize="small" /> },
+    { label: 'Refund Source', icon: <CurrencyExchangeIcon fontSize="small" /> },
+    { label: 'Expense Type', icon: <CurrencyExchangeIcon fontSize="small" /> }
+  ];
+
+  // grab all unique refund source names
+  const refundSourceNames: string[] = Array.from(
+    new Set(
+      reimbursementRequest.reimbursementProducts.flatMap((product) =>
+        product.refundSources.map((rs) => rs.indexCode.code + ' ' + rs.indexCode.name)
```
Comment (wavehassman) [inline src/frontend/src/pages/FinancePage/ReimbursementRequestDetailPage/ReimbursementRequestDetailsView.tsx]: put a dash between code and name like mock

File: src/frontend/src/pages/FinancePage/ReimbursementRequestDetailPage/ReimbursementRequestDetailsView.tsx [hunk 8]
```diff
@@ -477,60 +464,139 @@ const ReimbursementRequestDetailsView: React.FC<ReimbursementRequestDetailsViewP
...
+          </Box>
+        </Box>
+        <Box
+          sx={{
+            borderBottom: '2px solid white',
+            mt: 2,
+            mb: 2,
+            width: 'calc(100% - 40px)'
+          }}
+        />
```
Comment (wavehassman) [inline src/frontend/src/pages/FinancePage/ReimbursementRequestDetailPage/ReimbursementRequestDetailsView.tsx:566]: We had a reimbursement request timeline merged in, so this should be added to the view as well

File: src/shared/src/types/reimbursement-requests-types.ts [hunk 9]
```diff
@@ -111,21 +118,23 @@ export interface ReimbursementProductCreateArgs {
...
 export interface ReimbursementProductFormArgs extends ReimbursementProductCreateArgs {
   reason: WbsNumber | OtherProductReason;
-  firstSourceAmount?: number;
-  secondSourceAmount?: number;
+  refundSources: RefundSource[];
 }
 
 export interface OtherReimbursementProductCreateArgs extends ReimbursementProductCreateArgs {
   reason: OtherProductReason;
+  refundSources: RefundSource[];
```
Comment (wavehassman) [inline src/shared/src/types/reimbursement-requests-types.ts]: since this extends ReimbursementProductCreateArgs, you shouldn't have to put it in each interface, same as below

**Review by wavehassman — CHANGES_REQUESTED** [review]
This is amazing!! For the create RR modal, they should only be allowed to use two so this should be fine for now. Also, the create RR modal should also be on the side page as well as the view RR modal.

File: src/frontend/src/pages/FinancePage/ReimbursementRequestDetailPage/ReimbursementRequestDetailsView.tsx (hunk 5 above)
Comment (JoshuaGoldberg) [reply src/frontend/src/pages/FinancePage/ReimbursementRequestDetailPage/ReimbursementRequestDetailsView.tsx]: Oh, I think I overlooked that the pill allowed for colors already. I can change it back.

File: src/frontend/src/apis/finance.api.ts (hunk 3 above)
Comment (JoshuaGoldberg) [reply src/frontend/src/apis/finance.api.ts]: Oops, forgot to delete that!

File: src/frontend/src/pages/FinancePage/ReimbursementRequestForm/ReimbursementProductTable.tsx [hunk 10]
```diff
@@ -46,6 +46,8 @@ interface ReimbursementProductTableProps {
   setValue: UseFormSetValue<ReimbursementRequestFormInput>;
   control: Control<ReimbursementRequestFormInput>;
   hasMultipleRefundSources?: boolean;
+  firstRefundSourceIndexCode?: IndexCode | undefined;
```
Comment (wavehassman) [inline src/frontend/src/pages/FinancePage/ReimbursementRequestForm/ReimbursementProductTable.tsx]: no need to have the Undefined here, same as second refund source

**Review by wavehassman — CHANGES_REQUESTED** [review]
Great job!! I left one very small comment, but when I checked out the branch I saw some bugs. The edit reimbursement request modal still takes up full screen, we want it to also be on side page popup. Also, for reimbursement request #1 there is an issue with the spacing on the timeline, it isn't in a straight line, and then for reimbursement request #2 the products section is getting all squished. I would add more seed data to be thorough about the look in all different situations.

File: src/shared/src/types/reimbursement-requests-types.ts [hunk 11]
```diff
@@ -103,16 +104,21 @@ export interface AccountCode {
   indexCodes: IndexCode[];
 }
 
+export interface RefundSource {
+  refundSourceId?: string;
```
Comment (wavehassman) [inline src/shared/src/types/reimbursement-requests-types.ts]: this should not be optional

**Review by wavehassman — CHANGES_REQUESTED** [review]
This is amazing! I left one comment and then there are some changes to match the epic. The actions drop down on the reimbursements page should not have create reimbursement request instead use the button on the bottom left corner. The products section of the view rr modal should contain the project wbs number if there is one, so you need another column. It is in the second photo in your ticket for reference. Also, I would makethe  seed data match up cause it’s kinda confusing right now. Other than that, great job!!

File: src/frontend/src/layouts/Sidebar/Sidebar.tsx [hunk 12]
```diff
@@ -81,7 +81,7 @@ const Sidebar = ({ drawerOpen, setDrawerOpen, moveContent, setMoveContent, organ
         {
           name: 'Reimbursments',
```
Comment (walker-sean) [inline src/frontend/src/layouts/Sidebar/Sidebar.tsx]: I don't know how no-one caught this lmao can you fix the spelling here

Comment (walker-sean) [reply src/frontend/src/layouts/Sidebar/Sidebar.tsx]: This stops being highlighted when you switch to the All Requests tab

File: src/frontend/src/pages/FinancePage/FinanceComponents/ReimbursementRequestInfo.tsx [hunk 13]
```diff
@@ -182,6 +193,38 @@ const ReimbursementRequestInfo = ({
...
+  const [showCreateSidePage, setShowCreateSidePage] = useState(false);
+  const { isLoading: receiptsIsLoading, mutateAsync: uploadReceipts } = useUploadManyReceipts();
+
+  const openSidePage = () => {
+    setShowSidePage(true);
+  };
+
+  const closeSidePage = () => {
+    setShowSidePage(false);
+    onCloseSidePage?.();
```
Comment (walker-sean) [inline src/frontend/src/pages/FinancePage/FinanceComponents/ReimbursementRequestInfo.tsx]: Is this ever undefined?

File: src/frontend/src/pages/FinancePage/FinanceComponents/ReimbursementRequestInfo.tsx [hunk 14]
```diff
@@ -358,6 +414,18 @@ const ReimbursementRequestInfo = ({
...
       </Box>
+      <SidePage
+        showPage={showCreateSidePage}
+        handleClose={closeCreateSidePage}
+        title={sidePageTitle}
+        component={
+          sidePageTitle === 'Create Reimbursement Request' ? (
+            <ReimbursementRequestForm submitText="Submit" submitData={onSubmitCreate} previousPage={routes.FINANCE} />
+          ) : (
+            <Typography>This is a side page</Typography>
```
Comment (walker-sean) [inline src/frontend/src/pages/FinancePage/FinanceComponents/ReimbursementRequestInfo.tsx]: Remove this

**Review by walker-sean — CHANGES_REQUESTED** [review]
Looks amazing! 

A few more things: 
1. The buttons on the create/edit forms should close the sidepage rather than redirect to finance page
2. When the edit page is opened, close the info page behind it

File: src/frontend/src/layouts/Sidebar/Sidebar.tsx (hunk 12 above)
Comment (JoshuaGoldberg) [reply src/frontend/src/layouts/Sidebar/Sidebar.tsx]: Fixed the highlighting issue by simply redirecting '/reimbursement-requests' to '/reimbursement-requests/my-requests', since that should be the default tab regardless (due to the way full page tabs work). The redirect allows the sidebar option to remain highlighted by simply directing to '/reimbursement-requests', which encompasses both the my-requests and all-requests tabs of the table.

File: src/frontend/src/pages/FinancePage/FinanceComponents/ReimbursementRequestInfo.tsx (hunk 13 above)
Comment (JoshuaGoldberg) [reply src/frontend/src/pages/FinancePage/FinanceComponents/ReimbursementRequestInfo.tsx]: Technically in the original finance page having it be optional is nice, but since I'm assuming that page will be removed/repurposed anyway, you're right in that it shouldn't ever be undefined (changed it as a result).

**Review by walker-sean — APPROVED** [review]
